Introduce MVVM

ivinjabraham opened the pull request:
This PR address #139. I cannot foresee all the necessary steps at the moment but I will change this body when I can see the plan taking a more concrete form. Ideally, I will be able to integrate it without causing regressions i.e have the new architecture work alongside the old. But it is unclear how to go about that at the moment.

My local version of `clippy` (v0.1.88) seems to have added a new check and is now throwing errors for string interpolation syntax across the codebase. I haven't fixed them in this PR since it seems logically out of place. Though I think someone should raise an issue if they can confirm that it's not just on my machine.

ivinjabraham committed:
refactor: rename `app` module to `model` and `App` struct to `Model`

The `App` struct serves as the model in the MVC pattern. The proposed
MVVM refactor will continue using it as the primary Model. Rename `App`
to `Model` to better reflect it's purpose.

Signed-off-by: Ivin Joel Abraham <ivinjabraham@gmail.com>

ivinjabraham committed:
refactor: rename `CurrentScreen` enum to `View` to improve clarity

The `View` is a crucial part of MVVM that will handle the UI. Rename the
struct from `CurrentScreen`, which conveys logic, to simply `View` that
will reflect it's purpose in the architecture.

Signed-off-by: Ivin Joel Abraham <ivinjabraham@gmail.com>

ivinjabraham committed:
refactor: introduce ViewModel trait

The MVVM refactor will require a trait for viewmodels to establish a
reliable interface. Introduce a basic implementation of the trait with
no usage.

Signed-off-by: Ivin Joel Abraham <ivinjabraham@gmail.com>

ivinjabraham committed:
refactor: derive `Eq`, `Hash` and `Copy` for View

To use as a key in a hashmap, a type must implement `Eq` and `Hash`.
Additionally, `Copy` is required as well to be used in the .entry()
method of Hashmap.

Derive `Eq`, `Copy` and `Hash` on View to use it as a key in Hashmaps.

Signed-off-by: Ivin Joel Abraham <ivinjabraham@gmail.com>
Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

davidbtadokoro commented on src/viewmodels/mod.rs:
Sorry if I am getting this wrong, but from reading the description you've put in the `ViewModelState` the following is what I grasped.

A screen, say, "Bookmarked Patchsets" will be represented by:

1. A variant of `ViewModelState`, say `BookmarkedScreen`, which will have a type `BookmarkedScreenState` type that implements the trait `State`;
2. And a type `BookmarkedVM` which will implement the trait `ViewModel`.

So, the component driving the execution resolves which `View` is the current screen (in this example, `BookmarkedPatchsets`) and will:

1. Display the UI to the user
2. Capture the input/event
3. Pass as an argument to `BookmarkedVM::handle_key()`

At this moment, `BookmarkedVM::handle_key()` (possibly) mutates `BookmarkedScreenState`, effectively updating it, and the component driving the execution can recover the state of the screen, change or overwrite the current screen.

I am close?

davidbtadokoro commented on src/viewmodels/mod.rs:
```diff
@@ -0,0 +1,36 @@
+    fn state(&self) -> ViewModelState;
+}
+
+/// `ViewModelState` is an enum intended to represent the various possible
+/// state structs used by different screens in the application. Each variant
+/// of this enum should wrap the state struct corresponding to a specific screen,
+/// allowing for type-safe handling and dynamic dispatch of view model states.
```
Maybe this is due to my lack of practical experience with MVVM, but I don't think I quite understood what you meant here (as you can see by my previous comment). I am not saying this is bad, just asking if you can explain it better for me :sweat_smile:

ivinjabraham committed:
feat(View): add stub implementation of draw_screen

The View component in MVVM handles UI layout and construction. Though it
is currently in the model directory, the `View` enum will be our View
component. Future commits will move this to it's own directory.

Add a stub implementation of a method that can draw the current
screen.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
fix(ViewModel): change handle_key parameter to handle KeyEvent

The previous implementation, `key_handling`, used to match `KeyEvent`
and not `Event`. Correct the mistake in the new version.

Additionally, remove the comments that were getting outdated quickly due
to rapid development.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
feat(App): introduce a central manager

The multiple components in the proposed MVVM architecture require a
central manager to orchestrate event handling, controlling and
orchestration. Introduce an `App` to be the central manager with draft
implementations for methods.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
refactor(View): rename `ui` directory to `view`

The `ui` directory contains the functions that build the screen which
belongs to the View component in the MVVM architecture.

Replace the old name with `views`.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
feat(View): introduce State parameter for `draw_screen`

Drawing a screen would require information about the screen's state,
which shouldn't be available in the View itself.

Pass state to the `draw_screen` function.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
refactor(View): move View implementation to View directory

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
refactor(View): rename `MailingListSelection` to `MailingLists`

The `View` variant `MailingListSelection` is slightly unconventional as
other variants that include a similar selection list such as
`LatestPatchsets` does not emphasize on the "Selection" aspect of the
View.

Rename the variant to better match with the others.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
refactor(View): pass `state` as reference to `draw_screen`

`draw_screen` should only be reading the state when rendering and any
changes should be done through the viewmodel i.e ownership should belong
to the viewmodel.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
feat(View): pass terminal reference to draw_screen

The ratatui Terminal instance is ultimately required to draw on the
screen. Hence, pass it to the View when required.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>

ivinjabraham committed:
feat(View): implement View for MailingLists

Port the previous MVC architecture's View of MailingList to the new MVVM
architecture.  Keep previous implementations to ensure `patch-hub` does
not break.

This implementation is not currently used, it is being introduced ahead
of main loop integration to avoid breaking compilation during the
transition.

Signed-off-by: Ivin Joel Abraham <ivinjoelabraham@proton.me>
